### PR TITLE
Reflection engine issue780 find generic type

### LINKED ISSUE #780: Reflection_Engine: Provide a way to find the lowest common shared type between a list of types
Given a `BH.oM.Geometry.Point` and a `BH.oM.Geometry.Polyline`, I expect to get `BH.oM.Geometry.IGeometry`

### PR DESCRIPTION
**Pull request by epignatelli**
### Issues addressed by this PR

Closes #780 

Accepts a list of types and returns the closest parent type to them all.

### Test files

https://burohappold.sharepoint.com/:u:/s/BHoM/EVYkpCYXqI5GkqiUoq8AqfAB_NN2AHR9mVa5ghMMwZAZQg?e=7moUiu

### Additional comments

I am not sure this code is required. We may find other solutions within the `System.Reflection` namespace that are shorter and do not require packing in a method. Monitoring https://github.com/BHoM/BHoM_UI/pull/69 will solve this.

### CODE REVIEW

File: Reflection_Engine/Query/ClassHierarchy.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        public static List<Type> ClassHierarchy(this Type type)
+        {
+            if (type == null)
+                return new List<Type>();
+
+            List<Type> classHierarchy = new List<Type>();
+            for (Type current = type; current != null; current = current.GetInterfaces().FirstOrDefault() ?? current.BaseType)
```
Comment (adecler) [inline Reflection_Engine/Query/ClassHierarchy.cs]: The `FirstOrDefault` is incorrect here. You need to collect all of them (plus base type) and then apply `ClassHierarchy` recursively on those for it to give the correct result.

Comment (epignatelli) [reply Reflection_Engine/Query/ClassHierarchy.cs]: I found `InheritedTypes` doing exactly what I need, with a small fix. I will delete `ClassHierarchy`

File: Reflection_Engine/Query/CommonBaseType.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                return typeof(object);
+
+            types = types.Where(t => t != null).ToList();
+            if (types.Count == 0)
+                return typeof(object);
+
+            List<List<Type>> inheritanceChain = new List<List<Type>>();
+
+            foreach (Type type in types.Distinct())
+                inheritanceChain.Add(type.InheritedTypes());
```
Comment (IsakNaslundBh) [inline Reflection_Engine/Query/CommonBaseType.cs]: `type.InheritedTypes()` does not cover the actual type itself, so you probably want to add that to the list. 

Then, for the inherited types, not sure if it is getting out all the base types. If you for example are working on type `C` where `C : B` and `B : A`, I am not sure if  `A` will be added to the list.

Comment (epignatelli) [reply Reflection_Engine/Query/CommonBaseType.cs]: Yes, agreed. If the BHoM_UI pr is ok with the latest changes (does not rely on this anymore) I think I am going to close this without merging - this was a bad rush.

**Comment by epignatelli** [comment]
Closing this, since it is not critical and can be solved later on. 
Also referencing https://github.com/BHoM/BHoM_UI/issues/72 (thanks @IsakNaslundBh ) as direct application of #780

**Pull request closed without merging**